### ISSUE TITLE
[bug] Auto-detection of CMAKE_SYSTEM_NAME not working for all cross-compilation scenarios

### ISSUE THREAD

**Issue by jasal82** [conan-io/conan#7826]
Conan has an auto-detection feature for `CMAKE_SYSTEM_NAME` which is enabled by default if a value is not set explicitly. I found the method `_cmake_cross_build_defines()` in `CMakeDefinitionsBuilder` which seems to perform that auto-detection:

```python
        # System name and system version
        if cmake_system_name is not True:  # String not empty
            definitions["CMAKE_SYSTEM_NAME"] = cmake_system_name
        else:  # detect if we are cross building and the system name and version
            if cross_building(self._conanfile):  # We are cross building
                if os_ != os_build:
                    if os_:  # the_os is the host (regular setting)
                        definitions["CMAKE_SYSTEM_NAME"] = {"Macos": "Darwin",
                                                            "iOS": "Darwin",
                                                            "tvOS": "Darwin",
                                                            "watchOS": "Darwin",
                                                            "Neutrino": "QNX"}.get(os_, os_)
                    else:
                        definitions["CMAKE_SYSTEM_NAME"] = "Generic"
```

So if `os_ == os_build` the variable is never set which is clearly wrong. A typical cross-toolchain for arm would have `os = Linux` and `os_build = Linux` but `arch` and `arch_build` are different. My suggestion would be to add another `else:` block on the `if os_ != os_build` level that also sets `CMAKE_SYSTEM_NAME` to `Generic`.

**Comment by memsharded** [comment]
Just to be sure are you suggesting?

```python
         if cross_building(self._conanfile):  # We are cross building
                if os_ != os_build:
                    if os_:  # the_os is the host (regular setting)
                        definitions["CMAKE_SYSTEM_NAME"] = {"Macos": "Darwin",
                                                            "iOS": "Darwin",
                                                            "tvOS": "Darwin",
                                                            "watchOS": "Darwin",
                                                            "Neutrino": "QNX"}.get(os_, os_)
                    else:
                        definitions["CMAKE_SYSTEM_NAME"] = "Generic"
          else:
                definitions["CMAKE_SYSTEM_NAME"] = "Generic"
```

Or why not something like:

```python
         if cross_building(self._conanfile):  # We are cross building
                if os_ != os_build and arch != arch_build:
                    if os_:  # the_os is the host (regular setting)
                        definitions["CMAKE_SYSTEM_NAME"] = {"Macos": "Darwin",
                                                            "iOS": "Darwin",
                                                            "tvOS": "Darwin",
                                                            "watchOS": "Darwin",
                                                            "Neutrino": "QNX"}.get(os_, os_)
                    else:
                        definitions["CMAKE_SYSTEM_NAME"] = "Generic"
```

It seems that CMAKE_SYSTEM_NAME should not be set unless you are cross-building, so I would say the second.

**Comment by jasal82** [comment]
Ah I'm sorry, that was wrong. What I meant was

```python
         if cross_building(self._conanfile):  # We are cross building
            if os_ != os_build:
                if os_:  # the_os is the host (regular setting)
                    definitions["CMAKE_SYSTEM_NAME"] = {"Macos": "Darwin",
                                                        "iOS": "Darwin",
                                                        "tvOS": "Darwin",
                                                        "watchOS": "Darwin",
                                                        "Neutrino": "QNX"}.get(os_, os_)
                else:
                    definitions["CMAKE_SYSTEM_NAME"] = "Generic"
            else:
                definitions["CMAKE_SYSTEM_NAME"] = "Generic"
```

I'll fix the description. I'd say that the solution with `if os_ != os_build or arch != arch_build` would technically be the same because that's probably what `cross_building()` is checking for, isn't it?

**Comment by memsharded** [comment]
Ok, that makes more sense, indeed. Pushed review.

**Comment by jgsogo** [comment]
Hi, @jasal82 

According to [CMake docs](https://cmake.org/cmake/help/v3.4/manual/cmake-toolchains.7.html#cross-compiling-for-linux), cross-compiling to `Linux ARM` should use `set(CMAKE_SYSTEM_NAME Linux)`, why are you suggesting to assign `Generic` to it? According to [this link](https://stackoverflow.com/questions/24907916/cmake-how-to-properly-setup-new-toolchain-to-avoid-system-is-unknown-to-cmake), using `Generic` is basically a way to avoid a warning (or not to use the CMake distributed files).

**Comment by jasal82** [comment]
@jgsogo If you can set `Linux`, even better, but it doesn't matter for my specific problem because any value in `CMAKE_SYSTEM_NAME` will enable the cross-compilation logic in CMake. In the current state it is completely broken, so I was suggesting the minimal viable fix. Feel free to adjust. The correct approach would probably be to map the value of `os_` to the available CMake platform names.

**Comment by jgsogo** [comment]
Thanks for the answer, I was confused about it. 👍

**Comment by jgsogo** [comment]
#7827 merged to `1.30.1`. It will add the `CMAKE_SYSTEM_NAME` when cross-compiling to `arm` using a `Linux` machine to build.

Thanks for reporting the issue.

**Comment by memsharded** [comment]
Closed by https://github.com/conan-io/conan/pull/7827
